### PR TITLE
new anchoredMC.sh for nowadays productions

### PR DESCRIPTION
**Pull request by catalinristea**
Ciao,

Please review and amend if found necessary.

Thanks, Catalin.

### CODE REVIEW

File: MC/run/ANCHOR/anchorMC.sh
Comment (chiarazampolli) [inline MC/run/ANCHOR/anchorMC.sh:131]: Hello @sawenzel , @catalinristea , 
Do we still need these?

Comment (chiarazampolli) [reply MC/run/ANCHOR/anchorMC.sh:131]: Note also that the list contains non-existing objects, 
```
[ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch///FT0/Calibration/ChannelTimeOffset/1686209727558/
[ERROR] Unable to find object /FT0/Calibration/ChannelTimeOffset/1686209727558, Aborting
[ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch///FV0/Calibration/ChannelTimeOffset/1686209727558/
[ERROR] Unable to find object /FV0/Calibration/ChannelTimeOffset/1686209727558, Aborting
[ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch///GLO/GRP/BunchFilling/1686209727558/
[ERROR] Unable to find object /GLO/GRP/BunchFilling/1686209727558, Aborting
```

Comment (sawenzel) [reply MC/run/ANCHOR/anchorMC.sh:131]: These are no longer needed.

File: MC/run/ANCHOR/anchorMC.sh
Comment (sawenzel) [inline MC/run/ANCHOR/anchorMC.sh:34]: shouldn't this list be dynamic and come from the data async scripts?
O2DPG already determines the list of detectors (somewhat) dynamically

File: MC/run/ANCHOR/anchorMC.sh
Comment (sawenzel) [inline MC/run/ANCHOR/anchorMC.sh:55]: @chiarazampolli : is this still needed?

File: MC/run/ANCHOR/anchorMC.sh
Comment (sawenzel) [inline MC/run/ANCHOR/anchorMC.sh:149]: what is this doing??

File: MC/run/ANCHOR/anchorMC.sh
Comment (chiarazampolli) [reply MC/run/ANCHOR/anchorMC.sh:55]: I don't know, maybe @wiechula does? Otherwise I need to check.

File: MC/run/ANCHOR/anchorMC.sh
Comment (chiarazampolli) [reply MC/run/ANCHOR/anchorMC.sh:149]: To avoid overlaps in ITS and MFT. What was observed is that with overlaps hits are missing, see https://indico.cern.ch/event/1339452/contributions/5638180/attachments/2748989/4784254/SCD_11_2023_ITSreport_mconcas.pdf

File: MC/run/ANCHOR/anchorMC.sh
Comment (noferini) [reply MC/run/ANCHOR/anchorMC.sh:55]: I think this is a way to replace data TPC dEdx parameterizations (splines) wtht different ones tuned on MC.
Energy loss in MC is not exactly the same as in data  and gain calibration are not yet ready for MC (if I remember correctly)

Comment (wiechula) [reply MC/run/ANCHOR/anchorMC.sh:55]: Is the setenv_extra.sh the default one from asyncReco https://github.com/AliceO2Group/O2DPG/blob/master/DATA/production/configurations/asyncReco/setenv_extra.sh ?

Comment (benedikt-voelkel) [reply MC/run/ANCHOR/anchorMC.sh:55]: Yes, should be: https://github.com/AliceO2Group/O2DPG/pull/1340/files#diff-ec62e7d374a5f942616197c7eb143f6ad3faba56988ed36c8f0f0e5b49f936e1R48

File: MC/run/ANCHOR/anchorMC.sh
Comment (sawenzel) [reply MC/run/ANCHOR/anchorMC.sh:149]: but I think this can be simply done by pre-fetching trivial CCDB alignment objects for ITS and MFT. I am wondering more why we need to call a special workflow executable for this.

**Comment by chiarazampolli** [comment]
Hello @catalinristea , 
Do you think this PR could be updated with the comments above? I think they have not been taken, yet.
Then we could review it again.
Chiara

**Comment by benedikt-voelkel** [comment]
Are there any news on this?
It would be super helpful if that script could eventually go into O2DPG so there is no need of maintaining local versions.
In addition, it could help a lot to ensure more consistency across different SW builds, e.g. when we have 2023 vs. 2022 simulation SW setups.

File: MC/run/ANCHOR/anchorMC.sh
Comment (catalinristea) [reply MC/run/ANCHOR/anchorMC.sh]: Done, thanks!

**Comment by catalinristea** [comment]
Ciao @benedikt-voelkel, all

I've just pushed changes from your tests
https://gitlab.cern.ch/o2simulationtools/alice-dpg-tests/-/blob/main/reco-2022/20231122/anchorMC811.sh?ref_type=heads
I did just a small change: since SETENV is a keyword for (t)csh shells, I've renamed it to DPGSETENV, in case one is using any of those. Please have a look to make sure I haven't done mistakes.

Apart from these last minute changes, for what concerns me, this PR can be merged.

Thank you!

Yours, Catalin.

**Comment by catalinristea** [comment]
...made consistent "_alt" for base/remaining args with explicit set of energy & system.

Thanks,
Catalin.

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [inline MC/run/ANCHOR/anchorMC.sh:50]: The logic should be changed to something like
```bash
[[ ! -f setenv_extra.sh ]] && cp ${DPGSETENV} .
```
If `setenv_extra.sh` is there already, nothing needs to be done.
If not, it needs to be copied.

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [reply MC/run/ANCHOR/anchorMC.sh:56]: The original comment should still be finally clarified before merging.

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [inline MC/run/ANCHOR/anchorMC.sh:107]: This should not be there. I had it in order to test various different versions with the same `anchorMC.sh`

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [inline MC/run/ANCHOR/anchorMC.sh:115]: See my comment above.

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [inline MC/run/ANCHOR/anchorMC.sh:125]: See my comment above.

**Review by benedikt-voelkel — CHANGES_REQUESTED** [review]

**Comment by benedikt-voelkel** [comment]
@catalinristea @chiarazampolli @noferini @sawenzel 
I rebased the branch and added some updates. Let me know what you think.
This is the commit message:
```
* prevent from sourcing the script

* fix to find correct setenv_extra.sh

* stop immediately if running async_pass.sh fails
  --> removed the condiction on "NO_MC"

* add some comments

* remove test and development-related code
```

After a final test we should be able to proceed with merging it.

File: MC/run/ANCHOR/anchorMC.sh
Comment (benedikt-voelkel) [inline MC/run/ANCHOR/anchorMC.sh:164]: I defined `ALICEO2_CCDB_LOCALCACHE` above explicitly now.
The reasons why it might have worked before
1. `ALICEO2_CCDB_LOCALCACHE` was always set in the parent stop of this script
1. It actually failed but the workflow could still run, downloading everything when it was needed
1. It actually failed but the workflow could still run, files are not actually needed

**Comment by benedikt-voelkel** [comment]
I added a few more updates.
Now testing it, `pregencollcontext` complains with
```
the argument ('2932420022') for option '--first-orbit' is invalid
```
First of all, is that a reasonable number? Sorry for my ignorance

If so, we might need to change https://github.com/AliceO2Group/AliceO2/blob/dev/Steer/src/CollisionContextTool.cxx#L48 to `unsigned int` as it is then also used here: https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/common/include/CommonDataFormat/InteractionRecord.h#L44

Also the orbit limit is much more than just a simple `int` would hold: https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/CTP/include/DataFormatsCTP/Scalers.h#L139

@sawenzel @chiarazampolli

**Comment by chiarazampolli** [comment]
Hello @benedikt-voelkel , 

Yes, the max orbit should be unsigned int, see https://github.com/AliceO2Group/AliceO2/blob/652c819713a734b6a415ccda9e677ed562fa5620/Common/Constants/include/CommonConstants/LHCConstants.h#L38
```
constexpr unsigned int MaxNOrbits = 0xffffffff;
```
1 orbit is ~88.9 um, so the number above gives 260'764'266 s, which is anyway too large. Where does it come from?
Who is printing the error?

**Comment by chiarazampolli** [comment]
Apologies, I missed a conversion factor in https://github.com/AliceO2Group/O2DPG/pull/1340#issuecomment-1900492008: the first orbit 2932420022 gives:

2932420022 * 88.9 um = 260'692.1399558 s ~ 3days

So it is reasonable. But we should change the type to unsigned int. 

As discussed with @benedikt-voelkel , the test was run with 22o, for which the anchored MC 23k2b/c did not fail, because we did not use the precoll context.

I believe that what also changed between 2022 and 2023 is that from some point on the orbit reset was issued at every start of fill, and a fill never lasts more that 1 day. So in 2023 we never got out of the int range.

Chiara

**Comment by benedikt-voelkel** [comment]
PR in O2 is here: https://github.com/AliceO2Group/AliceO2/pull/12565

**Comment by chiarazampolli** [comment]
Hello @benedikt-voelkel , 
Thanks! I think we can anyway merge this, since the problem with the IR is there anyway.
Chiara

**Comment by lietava** [comment]
@chiarazampolli : clarifying the orbit reset 22 versus 23 with Anton Jusko in email you sent.

**Comment by benedikt-voelkel** [comment]
Let's wait for this one to be merged: https://github.com/AliceO2Group/O2DPG/pull/1416. Then, we can really merge this one here as well.

**Review by benedikt-voelkel — APPROVED** [review]

**Pull request merged**
